### PR TITLE
Add support for PHP 7.4

### LINKED ISSUE #9179: Add PHP 7.4 support
We've decided to use GitHub issues as a bug tracker only so the bug reports wouldn't be hidden in long list of ideas and support questions. Please go to https://forum.mautic.org/c/ideas to find support and discuss features.

since i can't sign up on the ideas site - something went wrong - i drop my question here. Any plans on becoming compatible with PHP 7.4.9 in the near future?

"Mautic does not support PHP version 7.4.9 at this time. To use Mautic, you will need to downgrade to an earlier version."

### PR DESCRIPTION
**Pull request by escopecz**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

~~Based on failing tests it looks like it will be related with https://github.com/mautic/mautic/issues/8288. If I understand it right, we won't be able to support PHP 7.3 and 7.4 at the same time~~

Adds support for PHP 7.4

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. 
2. 

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. 

#### List deprecations along with the new alternative:
1. 
2. 

#### List backwards compatibility breaks:
1. 
2.

### CODE REVIEW

**Comment by afonseca08** [comment]
Great to see this is in progress, what is needed to proceed with this?

**Comment by escopecz** [comment]
I'm actually stuck with this. As I explained in the description. Any suggestion is welcome.

**Comment by afonseca08** [comment]
Require PHP 7.4 as it’s the current stable version today and will have been so for a while by the time Mauric 3 is publicly released.

**Comment by escopecz** [comment]
Not that simple. We should do everything possible to support all PHP versions that get security patches. I can't even imagine how the Mautic upgrade would work if one version would require PHP 7.3 and the other 7.4.

**Comment by afonseca08** [comment]
I'm not familiar with the ocramius/proxy-manage library which does not support PHP 7.3 in the latest release. Is there an alternative that is viable or is it too deeply coupled in Mautic?

**Comment by escopecz** [comment]
I traced it down in #8288. It's dependency of the `doctrine/migrations` package. We need that.

**Comment by afonseca08** [comment]
If no other workaround, would we make users wait a year and a half (until Jan 2022) when 7.3 is no longer getting security patches to require 7.4 minimum? I'd hope not, so the question is when to support PHP 7.4 if it means ending Mautic support for 7.3 while not end of life?

**Comment by RCheesley** [comment]
This might be helpful: https://docs.google.com/spreadsheets/d/1dHEcJywlY_BQ4z7WtRPv7-VTUi-u1-xMuqc78AJUuwI/edit?usp=sharing there are a few dependencies that we will need to bump to support 7.4 - feel free to add any more detail!

**Comment by dennisameling** [comment]
Just resolved the conflicts and rebased to continue progress on this PR 🚀

**Comment by dennisameling** [comment]
@escopecz I think we're pretty much good to go with PHP 7.4 support now (looks like we can support 7.2 + 7.3 + 7.4 without problems), but we need to update Travis from `trusty` to `xenial` for PHP 7.4 support. The distro was forced to `trusty` [a few months ago](https://github.com/mautic/mautic/pull/7611) as it had [the exact same errors in Travis](https://travis-ci.org/github/mautic/mautic/jobs/546763160#L579) that [I'm facing now](https://travis-ci.org/github/mautic/mautic/jobs/709742292#L823). 

I've spent some hours trying to troubleshoot this. I have the feeling that some dependency is missing in Travis' [Xenial image](https://docs.travis-ci.com/user/trusty-to-xenial-migration-guide), but I can't figure out why. When I try to [debug the Travis build](https://docs.travis-ci.com/user/running-build-in-debug-mode/#restarting-a-job-in-debug-mode-via-api) I'm getting "access denied" as I think I don't have enough permissions in Travis to restart a job in debug mode 😞

The weird thing is that the tests mentioned in Travis succeed without errors on my local machine. So I'm 99% sure it's just a matter of a missing dependency or misconfiguration in Travis, not a problem with Mautic itself. 

@escopecz Do you have enough permissions in Travis to start a debug build of job `709742292`? 🤞

**Comment by alanhartless** [comment]
@dennisameling I restarted it. Will see what happens.

**Comment by dennisameling** [comment]
I was able to run a debug build in a separate Mautic repo (those debug builds only run for 30 mins in which you can SSH into the container). I compared trusty (Ubuntu 14.04) to xenial (Ubuntu 16.04); the PHP modules are exactly the same and the general config is almost the same. In the below ZIP you can find the outputs of `php -i` and `php -m` that I captured while in debug mode (both trusty and xenial).

[travis-php-modules.zip](https://github.com/mautic/mautic/files/4947666/travis-php-modules.zip)

Another difference between the two is that MySQL [was upgraded from 5.6 to 5.7 in Xenial](https://docs.travis-ci.com/user/database-setup/#mysql). Not sure if that makes any difference...

I captured the test log but it doesn't show any exceptions or more details:
[mautic_test.zip](https://github.com/mautic/mautic/files/4947712/mautic_test.zip)

Travis' [trusty to xenial migration guide](https://docs.travis-ci.com/user/trusty-to-xenial-migration-guide) also doesn't seem to provide any hints.

@escopecz do you have any pointers as to what this could be? The problem is not related to the changes in this PR but only to the upgrade to Xenial, which was attempted several months ago as well (see https://github.com/mautic/mautic/pull/7611).

**Comment by escopecz** [comment]
I have no clues. I asked a couple of team members who take care about dev environment and CI configuration for our team, but without any luck. 😞

**Comment by dennisameling** [comment]
PR blocked by https://github.com/mautic/mautic/issues/9072

**Comment by RCheesley** [comment]
Bumping to 3.2 to give us more space to figure this out! :)

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by afonseca08** [comment]
I've been following this issue for some time and was wondering what the latest update might be. It appears this is blocked due to the test suite itself failing, rather than an issue with the feature. Does anyone have a suggestion on how to unblock this issue?

**Comment by dennisameling** [comment]
@afonseca08 That's correct. We need to upgrade the test suite to use Ubuntu 16.04+, which adds support for PHP 7.4 in Travis. Planning to work on https://github.com/mautic/mautic/issues/9072 in the coming weekend. To be continued 👍

**Comment by dennisameling** [comment]
**Update**: this PR will be unblocked when https://github.com/mautic/mautic/pull/9266 is merged 🚀 🎉 if we can get that PR and this one merged for the 3.2 release (which I think is very much doable), then Mautic will start supporting PHP 7.4 on [November 23, 2020](https://github.com/mautic/mautic/milestone/66) 😄

**Comment by dennisameling** [comment]
This PR can be rebased as soon as the `3.1` branch is merged into `staging` again, I expect that to happen in the coming days. Looking forward to pushing this over the finish line 🚀

**Review by dennisameling — DISMISSED** [review]

**Comment by dennisameling** [comment]
@escopecz This should be good to merge now. Only thing is that this increases total Travis time because Travis can only run 4 jobs simultaneously, and PHP 7.4 adds another 2 jobs 😞 (https://travis-ci.org/github/mautic/mautic/builds/737174886). 

I'm planning to create a PR before the end of this month which switches to GitHub Actions, there we'll have 10 simultaneous jobs in 1 run and also multiple runs (e.g. multiple PRs that need to be tested simultaneously, Travis does them one after the other), so I suppose it's not too much of a problem for now. Do you have any thoughts on that?

**Review by dennisameling — DISMISSED** [review]

**Review by dennisameling — APPROVED** [review]

File: app/bundles/CoreBundle/Helper/PageHelper.php
Comment (escopecz) [inline app/bundles/CoreBundle/Helper/PageHelper.php:18]: I really like that you did this instead of doing it in each controller. However I don't get:
1. How is it related to PHP 7.4?
2. Why was this used only in some controllers?

**Review by escopecz — COMMENTED** [review]
Looks good! However it contains more changes than I expected.

File: app/bundles/CoreBundle/Helper/PageHelper.php
Comment (dennisameling) [reply app/bundles/CoreBundle/Helper/PageHelper.php:18]: I have no clue, this commit was done by you AFAIK :joy: https://github.com/mautic/mautic/pull/8447/commits/8ab8a2b6c35e37c4a0eea00924541dcb665486cb. I just built further on your already existing PR, thought these changes were done to properly support PHP 7.4. Probably this somehow got combined with another PR? Should we create a new PR you think?

Comment (escopecz) [reply app/bundles/CoreBundle/Helper/PageHelper.php:18]: Haha. OK. My poor memory. 🤦  So the question is why I did it :D Probably some small change I had to do in multiple controllers and decided to refactor it and make the change only in one place instead.

**Comment by escopecz** [comment]
I can't approve my own PR, but as Dennis finished it then I approve his changes by this comment 👍

**Comment by RCheesley** [comment]
Great - let's get this merged into 3.2!  Huge thanks to everybody who has helped with this, it may not be an exciting and shiny new feature but it really does help us to keep Mautic moving forward! 🎉

**Comment by dennisameling** [comment]
Merging now already so that new + rebased PRs for 3.2 will also be checked against PHP 7.4 👍

**Comment by frankofno** [comment]
actually, that's a very shiny forward move for me. Can't wait to see mautic supporting php 7.4 :)

**Pull request merged**
